src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_Nullable.cs

```diff
@@ -383,8 +446,41 @@ static void G(A a, B b)
     }
 }";
             var comp2 = CreateCompilation(source2, parseOptions: TestOptions.Regular8, references: new[] { comp.EmitToImageReference() });
-            // PROTOTYPE(NullableReferenceTypes): Should report warnings.
-            comp2.VerifyEmitDiagnostics();
+            comp2.VerifyDiagnostics(
+                // (8,14): warning CS8620: Nullability of reference types in argument of type 'A' doesn't match target type 'I<(object, object?)>' for parameter 'b' in 'void C.F(I<(object, object)> a, I<(object, object?)> b)'.
+                //         F(a, a);
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "a").WithArguments("A", "I<(object, object?)>", "b", "void C.F(I<(object, object)> a, I<(object, object?)> b)").WithLocation(8, 14),
+                // (9,11): warning CS8620: Nullability of reference types in argument of type 'B' doesn't match target type 'I<(object, object)>' for parameter 'a' in 'void C.F(I<(object, object)> a, I<(object, object?)> b)'.
+                //         F(b, b);
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInArgument, "b").WithArguments("B", "I<(object, object)>", "a", "void C.F(I<(object, object)> a, I<(object, object?)> b)").WithLocation(9, 11));
+
+            var type = comp2.GetMember<NamedTypeSymbol>("A");
+            Assert.Equal("I<(System.Object X, System.Object Y)>", type.Interfaces()[0].ToTestDisplayString());
+            type = comp2.GetMember<NamedTypeSymbol>("B");
+            Assert.Equal("I<(System.Object X, System.Object? Y)>", type.Interfaces()[0].ToTestDisplayString());
+        }
+
+        // PROTOTYPE(NullableReferenceTypes): Execute some of these same tests with feature disabled.
+
+        [Fact]
+        public void EmitAttribute_Constraint()
+        {
+            var source =
+@"public class A<T>
+{
+}
+public class B<T> where T : A<object?>
+{
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            CompileAndVerify(comp, validator: assembly =>
+            {
+                var reader = assembly.GetMetadataReader();
+                var typeDef = GetTypeDefinitionByName(reader, "B`1");
+                var typeParameter = reader.GetGenericParameter(typeDef.GetGenericParameters()[0]);
+                var constraint = reader.GetGenericParameterConstraint(typeParameter.GetConstraints()[0]);
+                AssertAttributes(reader, constraint.GetCustomAttributes(), "MethodDefinition:Void System.Runtime.CompilerServices.NullableAttribute..ctor(Boolean[])");
+            });
         }
```

Comment:
It would be nice to validate these through the symbol table as well.

Reply:
> It would be nice to validate these through the symbol table as well.

This validate source symbol, I believe @gafter meant to validate symbols for emitted metadata. CompileAndVerify takes a symbol validator delegate for that purpose.  

---
In reply to: [194556845](https://github.com/dotnet/roslyn/pull/27589#discussion_r194556845) [](ancestors = 194556845)